server/sonar-server/src/main/java/org/sonar/server/settings/ws/ValuesAction.java

```diff
@@ -0,0 +1,291 @@
+      if (definition != null && definition.multiValues()) {
+        valueBuilder.addAllValues(COMMA_SPLITTER.split(value));
+      } else {
+        valueBuilder.setValue(value);
+      }
+    }
+
+    private List<PropertyDto> loadProperties(DbSession dbSession, boolean loadGlobal, Set<String> keys) {
+      if (loadGlobal) {
+        return dbClient.propertiesDao().selectGlobalPropertiesByKeys(dbSession, keys);
+      }
+      return dbClient.propertiesDao().selectComponentPropertiesByKeys(dbSession, keys, component.getId());
+    }
+
+    private class PropertySetValues {
+      private final Map<String, PropertyKeyWithFieldAndSetId> propertyKeyWithFieldAndSetIds = new HashMap<>();
+      private final Map<String, PropertySetValue> propertySetValuesByPropertyKey = new HashMap<>();
+
+      PropertySetValues(List<PropertyDto> properties, boolean loadGlobal) {
+        properties.stream()
+          .filter(propertyDto -> propertySetKeys.contains(propertyDto.getKey()))
+          .forEach(propertyDto -> definitionsByKey.get(propertyDto.getKey()).fields()
+            .forEach(field -> COMMA_SPLITTER.splitToList(propertyDto.getValue())
+              .forEach(value -> add(propertyDto.getKey(), field.key(), value))));
+
+        loadProperties(dbSession, loadGlobal, propertyKeyWithFieldAndSetIds.keySet())
+          .forEach(propertySetDto -> {
+            PropertyKeyWithFieldAndSetId propertyKeyWithFieldAndSetIdKey = propertyKeyWithFieldAndSetIds.get(propertySetDto.getKey());
+            PropertySetValue propertySetValue = getOrCreatePropertySetValue(propertyKeyWithFieldAndSetIdKey.getPropertyKey());
+            propertySetValue.add(propertyKeyWithFieldAndSetIdKey.getSetId(), propertyKeyWithFieldAndSetIdKey.getFieldKey(), propertySetDto.getValue());
+          });
+      }
+
+      private void add(String propertyKey, String fieldKey, String value) {
+        String propertySetKey = generatePropertySetKey(propertyKey, value, fieldKey);
+        propertyKeyWithFieldAndSetIds.put(propertySetKey, new PropertyKeyWithFieldAndSetId(propertyKey, fieldKey, value));
+      }
+
+      private PropertySetValue getOrCreatePropertySetValue(String propertyKey) {
+        PropertySetValue propertySetValue = propertySetValuesByPropertyKey.get(propertyKey);
+        if (propertySetValue == null) {
+          propertySetValue = new PropertySetValue();
+          propertySetValuesByPropertyKey.put(propertyKey, propertySetValue);
+        }
+        return propertySetValue;
+      }
+
+      List<Map<String, String>> get(String propertyKey) {
+        return propertySetValuesByPropertyKey.get(propertyKey).get();
+      }
+    }
+  }
+
+  private List<PropertyDefinition> getDefinitions(Set<String> keys) {
+    return propertyDefinitions.getAll().stream()
+      .filter(def -> keys.contains(def.key()))
+      .collect(Collectors.toList());
+  }
+
+  private static String generatePropertySetKey(String propertyKey, String id, String fieldKey) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move this method into "ValuesBuilder". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3398)